Integration test fails with "timed out waiting for the condition"

<!--
Thank you for helping to improve Crossplane!

Please be sure to search for open issues before raising a new one. We use issues
for bug reports and feature requests. Please find us at https://slack.crossplane.io
for questions, support, and discussion.
-->

### What happened?
<!--
Please let us know what behaviour you expected and how Crossplane diverged from
that behaviour.
-->

Running the CI for a PR results in the following error:

```
>>>>>>> waiting for deployment crossplane rollout to finish
Waiting for deployment spec update to be observed...
Waiting for deployment spec update to be observed...
Waiting for deployment "crossplane" rollout to finish: 0 out of 1 new replicas have been updated...
Waiting for deployment "crossplane" rollout to finish: 0 of 1 updated replicas are available...
error: timed out waiting for the condition

>>>>>>> Cleaning up...
Deleting cluster "build-c17017a6-inttests" ...
22:23:35 [FAIL]
make[1]: *** [test-integration] Error 1
Makefile:93: recipe for target 'test-integration' failed
build/makelib/common.mk:351: recipe for target 'e2e' failed
make: *** [e2e] Error 2
Error: Process completed with exit code 2.
```

### How can we reproduce it?
<!--
Help us to reproduce your bug as succinctly and precisely as possible. Artifacts
such as example manifests or a script that triggers the issue are highly
appreciated!
-->

Run the CI

### What environment did it happen in?
Crossplane version:  1.4.0-rc.0.161.gee800943

<!--
Include at least the version or commit of Crossplane you were running. Consider
also including your:

* Cloud provider or hardware configuration
* Kubernetes version (use `kubectl version`)
* Kubernetes distribution (e.g. Tectonic, GKE, OpenShift)
* OS (e.g. from /etc/os-release)
* Kernel (e.g. `uname -a`)
-->

--- a/Makefile
+++ b/Makefile
@@ -5,6 +5,11 @@ PROJECT_NAME := provider-ibm-cloud
 PROJECT_REPO := github.com/crossplane-contrib/$(PROJECT_NAME)
 
 PLATFORMS ?= linux_amd64 linux_arm64
+
+# kind-related versions
+KIND_VERSION ?= v0.11.1
+KIND_NODE_IMAGE_TAG ?= v1.19.11
+
 # -include will silently skip missing files, which allows us
 # to load those files with a target in the Makefile. If only
 # "include" was used, the make command would fail and refuse
@@ -91,7 +96,7 @@ e2e.run: test-integration
 # Run integration tests.
 test-integration: $(KIND) $(KUBECTL) $(HELM3)
 	@$(INFO) running integration tests using kind $(KIND_VERSION)
-	@$(ROOT_DIR)/cluster/local/integration_tests.sh || $(FAIL)
+	@KIND_NODE_IMAGE_TAG=${KIND_NODE_IMAGE_TAG} $(ROOT_DIR)/cluster/local/integration_tests.sh || $(FAIL)
 	@$(OK) integration tests passed
 
 # Update the submodules, such as the common build scripts.

--- a/cluster/local/integration_tests.sh
+++ b/cluster/local/integration_tests.sh
@@ -32,103 +32,6 @@ echo_error(){
     exit 1
 }
 
-# k8s watchers
-
-wait_for_pods_in_namespace(){
-    local timeout=$1
-    shift
-    namespace=$1
-    shift
-    arr=("$@")
-    local counter=0
-    for i in "${arr[@]}";
-        do
-            echo -n "waiting for pod $i in namespace $namespace..." >&2
-            while ! ("${KUBECTL}" -n $namespace get pod $i) &>/dev/null; do
-                if [ "$counter" -ge "$timeout" ]; then echo "TIMEOUT"; exit -1; else (( counter+=5 )); fi
-                echo -n "." >&2
-                sleep 5
-            done
-            echo "FOUND POD!" >&2
-        done
-}
-
-check_deployments(){
-    for name in $1; do
-        echo_sub_step "inspecting deployment '${name}'"
-        local dep_stat=$("${KUBECTL}" -n "$2" get deployments/"${name}")
-
-        echo_info "check if is deployed"
-        if $(echo "$dep_stat" | grep -iq 'No resources found'); then
-            echo "is not deployed"
-            exit -1
-        else
-            echo_step_completed
-        fi
-
-        echo_info "check if is ready"
-        IFS='/' read -ra ready_status_parts <<< "$(echo "$dep_stat" | awk ' FNR > 1 {print $2}')"
-        if (("${ready_status_parts[0]}" < "${ready_status_parts[1]}")); then
-            echo "is not Ready"
-            exit -1
-        else
-            echo_step_completed
-        fi
-        echo
-    done
-}
-
-check_pods(){
-    pods=$("${KUBECTL}" -n "${CROSSPLANE_NAMESPACE}" get pods)
-    count=$(echo "$pods" | wc -l)
-    if (("${count}"-1 != "${1}")); then
-        sleep 10
-        "${KUBECTL}" get events -A
-        sleep 20
-        echo_error "unexpected number of pods"
-        exit -1
-    fi
-    echo "$pods"
-    while read -r pod_stat; do
-        name=$(echo "$pod_stat" | awk '{print $1}')
-        echo_sub_step "inspecting pod '${name}'"
-
-        if $(echo "$pod_stat" | awk '{print $3}' | grep -ivq 'Completed'); then
-            echo_info "is not completed, continuing with further checks"
-        else
-            echo_info "is completed, foregoing further checks"
-            echo_step_completed
-            continue
-        fi
-
-        echo_info "check if is ready"
-        IFS='/' read -ra ready_status_parts <<< "$(echo "$pod_stat" | awk '{print $2}')"
-        if (("${ready_status_parts[0]}" < "${ready_status_parts[1]}")); then
-            echo_error "is not ready"
-            exit -1
-        else
-            echo_step_completed
-        fi
-
-        echo_info "check if is running"
-        if $(echo "$pod_stat" | awk '{print $3}' | grep -ivq 'Running'); then
-            echo_error "is not running"
-            exit -1
-        else
-            echo_step_completed
-        fi
-
-        echo_info "check if has restarts"
-        if (( $(echo "$pod_stat" | awk '{print $4}') > 0 )); then
-            echo_error "has restarts"
-            exit -1
-        else
-            echo_step_completed
-        fi
-        echo
-    done <<< "$(echo "$pods" | awk 'FNR>1')"
-}
-
 # ------------------------------
 projectdir="$( cd "$( dirname "${BASH_SOURCE[0]}")"/../.. && pwd )"
 
@@ -137,9 +40,9 @@ eval $(make --no-print-directory -C ${projectdir} build.vars)
 
 # ------------------------------
 
-HOSTARCH="${HOSTARCH:-amd64}"
-BUILD_IMAGE="${BUILD_REGISTRY}/${PROJECT_NAME}-${HOSTARCH}"
-CONTROLLER_IMAGE="${BUILD_REGISTRY}/${PROJECT_NAME}-controller-${HOSTARCH}"
+SAFEHOSTARCH="${SAFEHOSTARCH:-amd64}"
+BUILD_IMAGE="${BUILD_REGISTRY}/${PROJECT_NAME}-${SAFEHOSTARCH}"
+CONTROLLER_IMAGE="${BUILD_REGISTRY}/${PROJECT_NAME}-controller-${SAFEHOSTARCH}"
 
 version_tag="$(cat ${projectdir}/_output/version)"
 # tag as latest version to load into kind cluster
@@ -168,7 +71,8 @@ echo "created cache dir at ${CACHE_PATH}"
 docker save "${BUILD_IMAGE}" -o "${CACHE_PATH}/${PACKAGE_NAME}.xpkg" && chmod 644 "${CACHE_PATH}/${PACKAGE_NAME}.xpkg"
 
 # create kind cluster with extra mounts
-echo_step "creating k8s cluster using kind"
+KIND_NODE_IMAGE="kindest/node:${KIND_NODE_IMAGE_TAG}"
+echo_step "creating k8s cluster using kind ${KIND_VERSION} and node image ${KIND_NODE_IMAGE}"
 KIND_CONFIG="$( cat <<EOF
 kind: Cluster
 apiVersion: kind.x-k8s.io/v1alpha4
@@ -179,16 +83,11 @@ nodes:
     containerPath: /cache
 EOF
 )"
-echo "${KIND_CONFIG}" | "${KIND}" create cluster --name="${K8S_CLUSTER}" --config=-
+echo "${KIND_CONFIG}" | "${KIND}" create cluster --name="${K8S_CLUSTER}" --wait=5m --image="${KIND_NODE_IMAGE}" --config=-
 
 # tag controller image and load it into kind cluster
 docker tag "${CONTROLLER_IMAGE}" "${PACKAGE_CONTROLLER_IMAGE}"
 "${KIND}" load docker-image "${PACKAGE_CONTROLLER_IMAGE}" --name="${K8S_CLUSTER}"
-
-# wait for kind pods
-echo_step "wait for kind pods"
-kindpods=("kube-apiserver-${BUILD_REGISTRY}-inttests-control-plane" "kube-controller-manager-${BUILD_REGISTRY}-inttests-control-plane" "kube-scheduler-${BUILD_REGISTRY}-inttests-control-plane")
-wait_for_pods_in_namespace 120 "kube-system" "${kindpods[@]}"
 
 # files are not synced properly from host to kind node container on Jenkins, so
 # we must manually copy image from host to node
@@ -244,28 +143,10 @@ echo_info "using crossplane version ${chart_version}"
 echo
 # we replace empty dir with our PVC so that the /cache dir in the kind node
 # container is exposed to the crossplane pod
-"${HELM3}" install crossplane --namespace crossplane-system crossplane-master/crossplane --version ${chart_version} --devel --set packageCache.pvc=package-cache
-
-echo_step "waiting for deployment crossplane rollout to finish"
-"${KUBECTL}" -n "${CROSSPLANE_NAMESPACE}" rollout status "deploy/crossplane" --timeout=2m
-
-echo_step "wait until the pods are up and running"
-"${KUBECTL}" -n "${CROSSPLANE_NAMESPACE}" wait --for=condition=Ready pods --all --timeout=1m
+"${HELM3}" install crossplane --namespace crossplane-system crossplane-master/crossplane --version ${chart_version} --devel --wait --set packageCache.pvc=package-cache
 
 # ----------- integration tests
 echo_step "--- INTEGRATION TESTS ---"
-echo
-echo_step "check for necessary deployment statuses"
-echo
-echo -------- deployments
-"${KUBECTL}" -n "${CROSSPLANE_NAMESPACE}" get deployments
-
-check_deployments "crossplane" "${CROSSPLANE_NAMESPACE}"
-
-echo_step "check for crossplane pods statuses"
-echo
-echo "--- pods ---"
-check_pods 2
 
 # install package
 echo_step "installing ${PROJECT_NAME} into \"${CROSSPLANE_NAMESPACE}\" namespace"
@@ -283,29 +164,29 @@ EOF
 
 echo "${INSTALL_YAML}" | "${KUBECTL}" apply -f -
 
-"${KUBECTL}" -n "${CROSSPLANE_NAMESPACE}" get deployments
-
-# this is to let package manager unpack pods and start the controller.
-sleep 20
-
 # printing the cache dir contents can be useful for troubleshooting failures
 echo_step "check kind node cache dir contents"
 docker exec "${K8S_CLUSTER}-control-plane" ls -la /cache
 
-echo_step "check for package pod statuses"
-echo
-echo "--- pods ---"
-check_pods 3
+echo_step "waiting for provider to be installed"
+
+kubectl wait "provider.pkg.crossplane.io/${PACKAGE_NAME}" --for=condition=healthy --timeout=60s
 
 echo_step "uninstalling ${PROJECT_NAME}"
 
 echo "${INSTALL_YAML}" | "${KUBECTL}" delete -f -
 
 # check pods deleted
-sleep 30
-echo_step "check only crossplane pods remain"
-echo "--- pods ---"
-"${KUBECTL}" -n "${CROSSPLANE_NAMESPACE}" get pods
-check_pods 2
+timeout=60
+current=0
+step=3
+while [[ $(kubectl get providerrevision.pkg.crossplane.io -o name | wc -l) -gt 0 ]]; do
+  echo "waiting for provider to be deleted for another $step seconds"
+  current=$current+$step
+  if ! [[ $timeout > $current ]]; then
+    echo_error "timeout of ${timeout}s has been reached"
+  fi
+  sleep $step;
+done
 
 echo_success "Integration tests succeeded!"

